lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs

```diff
@@ -0,0 +1,529 @@
+        parseToken policy (NestedTokenQuantity token quantity) = do
+            when (TQ.isZero quantity) $
+                jsonFailWithZeroValueTokenQuantity policy token
+            pure (token, quantity)
+
+-- Used for JSON serialization only: not exported.
+data NestedMapEntry = NestedMapEntry
+    { _policy :: !TokenPolicyId
+    , _tokens :: ![NestedTokenQuantity]
+    } deriving Generic
+
+-- Used for JSON serialization only: not exported.
+data NestedTokenQuantity = NestedTokenQuantity
+    { _token :: !TokenName
+    , _quantity :: !TokenQuantity
+    } deriving Generic
+
+instance FromJSON NestedMapEntry where
+    parseJSON = genericParseJSON jsonOptions
+instance ToJSON NestedMapEntry where
+    toJSON = genericToJSON jsonOptions
+
+instance FromJSON NestedTokenQuantity where
+    parseJSON = genericParseJSON jsonOptions
+instance ToJSON NestedTokenQuantity where
+    toJSON = genericToJSON jsonOptions
+
+--------------------------------------------------------------------------------
+-- Construction
+--------------------------------------------------------------------------------
+
+-- | The empty token map.
+--
+empty :: TokenMap
+empty = TokenMap mempty
+
+-- | Creates a singleton token map with just one token quantity.
+--
+-- If the specified token quantity is zero, then the resultant map will be
+-- equal to the 'empty' map.
+--
+singleton :: AssetId -> TokenQuantity -> TokenMap
+singleton = setQuantity empty
+
+-- | Creates a token map from a flat list.
+--
+-- If a token name appears more than once in the list under the same policy,
+-- its associated quantities will be added together in the resultant map.
+--
+fromFlatList :: [(AssetId, TokenQuantity)] -> TokenMap
+fromFlatList = F.foldl' acc empty
+  where
+    acc b (asset, quantity) = adjustQuantity b asset (<> quantity)
+
+-- | Creates a token map from a nested list.
+--
+-- If a token name appears more than once in the list under the same policy,
+-- its associated quantities will be added together in the resultant map.
+--
+fromNestedList
+    :: [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))] -> TokenMap
+fromNestedList entries = fromFlatList
+    [ (AssetId policy token, quantity)
+    | (policy, tokenQuantities) <- entries
+    , (token, quantity) <- NE.toList tokenQuantities
+    ]
+
+--------------------------------------------------------------------------------
+-- Deconstruction
+--------------------------------------------------------------------------------
+
+-- | Converts a token map to a flat list.
+--
+toFlatList :: TokenMap -> [(AssetId, TokenQuantity)]
+toFlatList b =
+    [ (AssetId policy token, quantity)
+    | (policy, tokenQuantities) <- toNestedList b
+    , (token, quantity) <- NE.toList tokenQuantities
+    ]
+
+-- | Converts a token map to a nested list.
+--
+toNestedList
+    :: TokenMap -> [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))]
+toNestedList =
+    fmap (fmap NEMap.toList) . Map.toList . unTokenMap
```

Comment:
> Would you mind at all if I were to attempt this in a follow-up PR?

Nope, and it was more of a gratuitous remark / reminder. I don't think it'd matter much for token maps because maps would be rather small anyway (bounded by the transaction max size).